module-apps/apps-common/ApplicationCommon.cpp

```diff
@@ -530,56 +540,81 @@ namespace app
+        auto switchData = std::move(msg->getData());
+        if (switchData && switchData->ignoreCurrentWindowOnStack) {
+            windowsStack().pop();
+        }
+        auto anotherWindowOnTop = (not isCurrentWindow(msg->getWindowName())) and (not windowsStack().isEmpty());
+        if (anotherWindowOnTop) {
+            const auto closeReason = msg->getReason() == SwitchReason::PhoneLock
+                                         ? gui::Window::CloseReason::PhoneLock
+                                         : gui::Window::CloseReason::WindowSwitch;
+            getCurrentWindow()->onClose(closeReason);
+        }
 
-        if (windowsFactory.isRegistered(msg->getWindowName())) {
-            auto switchData = std::move(msg->getData());
-            if (switchData && switchData->ignoreCurrentWindowOnStack) {
-                popToWindow(getPrevWindow());
-            }
-            if (!isCurrentWindow(msg->getWindowName())) {
-                if (!windowsStack.isEmpty()) {
-                    const auto closeReason = msg->getReason() == SwitchReason::PhoneLock
-                                                 ? gui::Window::CloseReason::PhoneLock
-                                                 : gui::Window::CloseReason::WindowSwitch;
-                    getCurrentWindow()->onClose(closeReason);
-                }
-                setActiveWindow(msg->getWindowName());
-            }
-            LOG_DEBUG("Current window: %s vs %s", getCurrentWindow()->getName().c_str(), msg->getWindowName().c_str());
-            getCurrentWindow()->handleSwitchData(switchData.get());
+        LOG_DEBUG("Current window: %s vs %s", getCurrentWindow()->getName().c_str(), msg->getWindowName().c_str());
+        // THIS IS IN TEST!
+        const auto &[name, data] = msg->getSwitchData();
+        pushWindow(name, data);
+        getCurrentWindow()->handleSwitchData(switchData.get());
 
-            auto ret = dynamic_cast<gui::SwitchSpecialChar *>(switchData.get());
-            if (ret != nullptr && ret->type == gui::SwitchSpecialChar::Type::Response) {
-                auto text = dynamic_cast<gui::Text *>(getCurrentWindow()->getFocusItem());
-                if (text != nullptr) {
-                    text->addText(ret->getDescription());
-                    refreshWindow(gui::RefreshModes::GUI_REFRESH_FAST);
-                    return sys::msgHandled();
-                }
-            }
-            getCurrentWindow()->onBeforeShow(msg->getCommand(), switchData.get());
+        if (handleUpdateTextRefresh(switchData.get())) {
+            return sys::msgHandled();
+        }
+        getCurrentWindow()->onBeforeShow(msg->getCommand(), switchData.get());
+
+        /// This is kind of gentle hack on our stateless Application Window management switching
+        /// if we are not requesting for popup and we cant handle next popup - just refresh the window
+        /// else -> popup will refresh it for us
+        if (not tryShowPopup()) {
```

Comment:
i would try to make it more readable ? make function to check condition here ? and second to handle it ? so you will have:

`if(checkPopupToShow){
showPopup()}`

Reply:
why isn't it readable? if you successfully show popup -> there is no refresh : with no action in different branches here. Therefore it should be actually more readable this way ( as you do not have to analyze two branches in this function which is one of the hardest elements for analysis in windows handling)